mujoco_py/mjpid.pyx

```diff
@@ -0,0 +1,122 @@
+    NUM_USER_DATA_PER_ACT = 3,
+
+
+cdef mjtNum c_zero_gains(const mjModel* m, const mjData* d, int id) with gil:
+    return 0.0
+
+
+cdef mjtNum c_pid_bias(const mjModel* m, const mjData* d, int id) with gil:
+    cdef double dt_in_sec = m.opt.timestep
+
+    cdef double error = d.ctrl[id] - d.actuator_length[id]
+
+    cdef double Kp = m.actuator_user[id * m.nuser_actuator + PROPORTIONAL_GAIN]
+    cdef double error_deadband = m.actuator_user[id * m.nuser_actuator + ERROR_DEADBAND]
+    cdef double integral_max_clamp = m.actuator_user[id * m.nuser_actuator + INTEGRAL_MAX_CLAMP]
+    cdef double integral_time_const = m.actuator_user[id * m.nuser_actuator + INTEGRAL_TIME_CONSTANT]
+    cdef double derivative_gain_smoothing = \
+        m.actuator_user[id * m.nuser_actuator + DERIVATIVE_GAIN_SMOOTHING]
+    cdef double derivate_time_const = m.actuator_user[id * m.nuser_actuator + DERIVATIVE_TIME_CONSTANT]
+
+    cdef double corrective_effort_limit = fmax(
+            fabs(m.actuator_forcerange[id * 2]), fabs(m.actuator_forcerange[id * 2 + 1]))
+
+    if fabs(error) < error_deadband:
+        error = 0.0
+
+    integral_error = d.userdata[id * NUM_USER_DATA_PER_ACT + INTEGRAL_ERROR]
+    integral_error += error * dt_in_sec
+    integral_error = fmax(-integral_max_clamp, fmin(integral_max_clamp, integral_error))
+
+    last_error = d.userdata[id * NUM_USER_DATA_PER_ACT + LAST_ERROR]
+    cdef double derivative_error = (error - last_error) / dt_in_sec
+
+    derivative_error_last = d.userdata[id * NUM_USER_DATA_PER_ACT + DERIVATIVE_ERROR_LAST]
+
+    derivative_error = (1.0 - derivative_gain_smoothing) * derivative_error_last + \
+        derivative_gain_smoothing * derivative_error
+
+    cdef double integral_error_term = 0.0
+    if integral_time_const != 0:
+        integral_error_term = integral_error / integral_time_const
+
+    cdef double derivative_error_term = 0.0
+    if derivate_time_const != 0:
+        derivative_error_term = derivative_error / derivate_time_const
+
+    f = Kp * (error + integral_error_term + derivative_error_term)
+    # print(id, error, integral_error_term, derivative_error_term, derivative_error, dt_in_sec,
+    #    last_error, integral_error, derivative_error_last)
+
+    d.userdata[id * NUM_USER_DATA_PER_ACT + LAST_ERROR] = error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + DERIVATIVE_ERROR_LAST] = derivative_error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + INTEGRAL_ERROR] = integral_error
+    return fmax(-corrective_effort_limit, fmin(corrective_effort_limit, f))
+
+
+def set_pid_control(m, d):
+    global mjcb_act_gain
+    global mjcb_act_bias
+
+    if m.nuserdata < m.nu * NUM_USER_DATA_PER_ACT:
+        raise Exception('nuserdata is not set large enough to store PID internal states')
+
+    for i in range(m.nuserdata):
+        d.userdata[i] = 0.0
+
+    for i in range(m.nu):
+        m.actuator_gaintype[i] = const.GAIN_USER
+        m.actuator_biastype[i] = const.BIAS_USER
```

Comment:
I don't think we should override that manually for the actuators here. That should be specified in the XML file.

Reply:
if we set xml type to "user", it forces everyone to switch to new PID; we certainly could, I shall choose a setting to be exactly the same as previous position control.  

I was thinking about "opt in", only if user set_pid_control, to turn on the behavior

Reply:
I also would like it to be opt-in, but on the XML level - then we can have multiple hands, one with PID Controller, one without. We can have even these two hand in the same simulation. Or we can have kuka arm with P controllers and a shadow hand with PID controllers. The more things we allow to configure via XML the better for flexibility.